weed/util/http/client/http_client.go

```diff
@@ -0,0 +1,205 @@
+}
+
+func (httpClient *HTTPClient) Post(url, contentType string, body io.Reader) (resp *http.Response, err error) {
+	url, err = httpClient.NormalizeHttpScheme(url)
+	if err != nil {
+		return nil, err
+	}
+	return httpClient.Client.Post(url, contentType, body)
+}
+
+func (httpClient *HTTPClient) PostForm(url string, data url.Values) (resp *http.Response, err error) {
+	url, err = httpClient.NormalizeHttpScheme(url)
+	if err != nil {
+		return nil, err
+	}
+	return httpClient.Client.PostForm(url, data)
+}
+
+func (httpClient *HTTPClient) Head(url string) (resp *http.Response, err error) {
+	url, err = httpClient.NormalizeHttpScheme(url)
+	if err != nil {
+		return nil, err
+	}
+	return httpClient.Client.Head(url)
+}
+func (httpClient *HTTPClient) CloseIdleConnections() {
+	httpClient.Client.CloseIdleConnections()
+}
+
+func (httpClient *HTTPClient) GetClientTransport() *http.Transport {
+	return httpClient.Transport
+}
+
+func (httpClient *HTTPClient) GetHttpScheme() string {
+	if httpClient.expectHttpsScheme {
+		return "https"
+	}
+	return "http"
+}
+
+func (httpClient *HTTPClient) NormalizeHttpScheme(rawURL string) (string, error) {
+	expectedScheme := httpClient.GetHttpScheme()
+
+	if !(strings.HasPrefix(rawURL, "http://") || strings.HasPrefix(rawURL, "https://")) {
+		return expectedScheme + "://" + rawURL, nil
+	}
+
+	parsedURL, err := url.Parse(rawURL)
+	if err != nil {
+		return "", err
+	}
+
+	if expectedScheme != parsedURL.Scheme {
+		parsedURL.Scheme = expectedScheme
+	}
+	return parsedURL.String(), nil
+}
+
+func NewHttpClient(clientName ClientName, opts ...HttpClientOpt) (*HTTPClient, error) {
+	httpClient := HTTPClient{}
+	httpClient.expectHttpsScheme = CheckIsHttpsClientEnabled(clientName)
+	var tlsConfig *tls.Config = nil
+
+	if httpClient.expectHttpsScheme {
+		clientCertPair, err := GetClientCertPair(clientName)
+		if err != nil {
+			return nil, err
+		}
+
+		clientCaCert, clientCaCertName, err := GetClientCaCert(clientName)
+		if err != nil {
+			return nil, err
+		}
+
+		if clientCertPair != nil || len(clientCaCert) != 0 {
+			caCertPool, err := CreateHTTPClientCertPool(clientCaCert, clientCaCertName)
+			if err != nil {
+				return nil, err
+			}
+
+			tlsConfig = &tls.Config{
+				Certificates:       []tls.Certificate{},
+				RootCAs:            caCertPool,
+				InsecureSkipVerify: false,
+			}
+
+			if clientCertPair != nil {
+				tlsConfig.Certificates = append(tlsConfig.Certificates, *clientCertPair)
+			}
+		}
+	}
+
+	httpClient.Transport = &http.Transport{
+		MaxIdleConns:        1024,
+		MaxIdleConnsPerHost: 1024,
+		TLSClientConfig:     tlsConfig,
+	}
+	httpClient.Client = &http.Client{
+		Transport: httpClient.Transport,
+	}
+
+	for _, opt := range opts {
+		opt(&httpClient)
+	}
+	return &httpClient, nil
+}
+
+func GetStringOptionFromSecurityConfiguration(clientName ClientName, stringOptionName string) string {
+	onceLoadSecurityConfiguration.Do(func() {
+		util.LoadConfiguration("security", false)
```

Comment:
`util.LoadConfiguration("security", false)` is also called in many other places. This call may be duplicated, either though it is just one more time.

Reply:
Moved the loading of the security config to a separate function: https://github.com/seaweedfs/seaweedfs/pull/5766/commits/9a04357a743324298df615fa96e4a6a679ed754d

Reply:
https://github.com/seaweedfs/seaweedfs/commit/9a04357a743324298df615fa96e4a6a679ed754d is a refactoring without changing the execution. [My question](https://github.com/seaweedfs/seaweedfs/pull/5766#discussion_r1674259211) is still the same.

Reply:
@chrislusf, hello!

Should we ensure that the `sequrity` call is executed only once in this specific place, or do we need to replace the `sequrity` call throughout the entire code?

If we need to ensure that the `sequrity` call is executed only once in this specific place, then it is already happening as the `sequrity` call is wrapped in a `sync.Once`.

If we need to replace the `sequrity` call throughout the entire code with the `loadSecurityConfig` function, then I think this pull request would be overloaded with additional changes. It would be better to handle this in a separate PR.

Reply:
> If we need to replace the `sequrity` call throughout the entire code with the `loadSecurityConfig` function, then I think this pull request would be overloaded with additional changes. It would be better to handle this in a separate PR.

This is better and please send a PR for it.

Reply:
@chrislusf, hi!

Added a fix: https://github.com/seaweedfs/seaweedfs/pull/5766/commits/55a578089773350dd708fb83522d4ddeef34881e